sql/src/main/java/io/crate/operation/scalar/DateTruncFunction.java

```diff
@@ -21,76 +21,222 @@
     }
 
     @Override
     public Symbol normalizeSymbol(Function symbol) {
-        assert (symbol.arguments().size() == 2);
+        assert symbol.arguments().size() > 1 && symbol.arguments().size() < 4 : "Invalid number of arguments";
+
+        if (anyNonLiterals(symbol.arguments())) {
+            return symbol;
+        }
 
         Literal interval = (Literal) symbol.arguments().get(0);
-        isValidInterval(interval, symbol);
-        Symbol tsSymbol = symbol.arguments().get(1);
-
-        if (tsSymbol.symbolType().isValueSymbol()) {
-            return Literal.newLiteral(
-                    DataTypes.TIMESTAMP,
-                    evaluate((BytesRef)interval.value(), DataTypes.TIMESTAMP.value(((Input) tsSymbol).value()))
-            );
+        Literal tsSymbol;
+        Literal timezone;
+
+        if (symbol.arguments().size() == 2) {
+            timezone = DEFAULT_TZ_LITERAL;
+            tsSymbol = (Literal)symbol.arguments().get(1);
         } else {
-            if ( !tsSymbol.valueType().equals(DataTypes.TIMESTAMP)) {
-                throw new IllegalArgumentException(SymbolFormatter.format(
-                        "The argument \"%s\" given to the date_trunc function has an invalid data type",
-                        tsSymbol));
-            }
+            timezone = (Literal) symbol.arguments().get(1);
+            tsSymbol = (Literal)symbol.arguments().get(2);
         }
-        return symbol;
+
+        return Literal.newLiteral(
+                DataTypes.TIMESTAMP,
+                evaluate(new Input[]{interval, timezone, tsSymbol})
+        );
     }
 
-    private Long evaluate(BytesRef interval, Long value) {
+    @Override
+    public final Long evaluate(Input[] args) {
+        assert args.length > 1 && args.length < 4 : "Invalid number of arguments";
+        Object value;
+        BytesRef timeZone = DEFAULT_TZ_BYTES_REF;
+        if (args.length == 2) {
+            value = args[1].value();
+        } else {
+            timeZone = (BytesRef)args[1].value();
+            value = args[2].value();
+        }
         if (value == null) {
             return null;
         }
-        DateTimeUnit fieldParser = DATE_FIELD_PARSERS.get(interval);
-        assert fieldParser != null;
-        return truncate(fieldParser, value, DEFAULT_TZ);
+        if (tzRounding == null) {
+            BytesRef interval = (BytesRef)args[0].value();
+            return truncate(rounding(interval, timeZone), TimestampType.INSTANCE.value(value));
+        }
+        return truncate(tzRounding, TimestampType.INSTANCE.value(value));
     }
 
-    @Override
-    public final Long evaluate(Input[] args) {
-        assert args.length == 2;
-        assert args[0].value() != null;
-        assert args[0].value() instanceof BytesRef;
-        return evaluate((BytesRef)args[0].value(), (Long)args[1].value());
+    protected Rounding rounding(BytesRef interval, BytesRef timeZoneString) {
+        isValidInterval(interval);
```

Comment:
instead of using containsKey in isValidInterval the return value in the following line could simply be checked against null?